hrms/payroll/doctype/salary_slip/salary_slip.py

```diff
@@ -1260,63 +1282,245 @@ def eval_condition_and_formula(self, struct_row, data):
 			raise
 
 	def add_employee_benefits(self):
-		for struct_row in self._salary_structure_doc.get("earnings"):
-			if struct_row.is_flexible_benefit == 1:
-				if (
-					frappe.db.get_value(
-						"Salary Component",
-						struct_row.salary_component,
-						"pay_against_benefit_claim",
-						cache=True,
-					)
-					!= 1
-				):
-					benefit_component_amount = get_benefit_component_amount(
-						self.employee,
-						self.start_date,
-						self.end_date,
-						struct_row.salary_component,
-						self._salary_structure_doc,
-						self.payroll_frequency,
-						self.payroll_period,
-					)
-					if benefit_component_amount:
-						self.update_component_row(struct_row, benefit_component_amount, "earnings")
-				else:
-					benefit_claim_amount = get_benefit_claim_amount(
-						self.employee, self.start_date, self.end_date, struct_row.salary_component
-					)
-					if benefit_claim_amount:
-						self.update_component_row(struct_row, benefit_claim_amount, "earnings")
+		"""Fetch employee benefits from Salary Structure Assignment. Get amounts for accrual or payouts for each and add to salary slip"""
+		EmployeeBenefitDetail = frappe.qb.DocType("Employee Benefit Detail")
+		SalaryComponent = frappe.qb.DocType("Salary Component")
+		employee_benefits = (
+			frappe.qb.from_(EmployeeBenefitDetail)
+			.join(SalaryComponent)
+			.on(EmployeeBenefitDetail.salary_component == SalaryComponent.name)
+			.select(
+				EmployeeBenefitDetail.salary_component,
+				EmployeeBenefitDetail.amount.as_("yearly_amount"),
+				SalaryComponent.payout_method,
+				SalaryComponent.depends_on_payment_days,
+				SalaryComponent.round_to_the_nearest_integer,
+				SalaryComponent.final_cycle_accrual_payout,
+			)
+			.where(EmployeeBenefitDetail.parent == self._salary_structure_assignment.name)
+			.where(SalaryComponent.is_flexible_benefit == 1)
+			.where(
+				(SalaryComponent.payout_method == "Accrue and payout at end of payroll period")
+				| (SalaryComponent.payout_method == "Accrue per cycle, pay only on claim")
+			)
```

Comment:
Can check for **Is Accrual Component** checkbox instead of this? The comparison maybe faster